common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java

```diff
@@ -106,28 +127,32 @@ public MaterialConfig config() {
     }
 
     public List<Modification> modificationsSince(File baseDir, Revision revision, final SubprocessExecutionContext execCtx) {
+        unshallowIfNeeded(ProcessOutputStreamConsumer.inMemoryConsumer(), revision, baseDir);
         GitCommand gitCommand = getGit(baseDir);
         return gitCommand.modificationsSince(revision);
     }
 
     public MaterialInstance createMaterialInstance() {
-        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString());
+        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString(), shallowClone);
     }
 
     @Override
     protected void appendCriteria(Map<String, Object> parameters) {
         parameters.put(ScmMaterialConfig.URL, url.forCommandline());
         parameters.put("branch", branch);
+        parameters.put("shallowClone", shallowClone);
```

Comment:
About material identity: My guess about submoduleFolder is that it's not even used. Maybe I'm wrong.

> I totally agree. I feel in an ideal implementations, flyweight clone on go server should always be shallow (because historical revision is never used), and shallow flag is turn on and off for agents. But that would involve too much changes to the existing process.

[This](https://github.com/gocd/gocd/blob/eb0ae575fbb0a30f59ed646cd45d16eb57d26aa5/server/src/com/thoughtworks/go/server/service/materials/GitPoller.java#L31) is the only place called from the MDU, for git material clone. Do you think it's worth trying to shallow that up? :) Behind a flag? If what I'm saying is right and it's only that place, it might not be too much of a change.